### PR TITLE
[CALCITE-4106] Consider "listCoerced" when TypeCoercionImpl#inOperationCoercion returns (Jiatao Tao)

### PR DESCRIPTION
**Pull request by Aaaaaaron**

### CODE REVIEW

**Comment by Aaaaaaron** [comment]
Hard to add UT.
testInOperation/testNotInOperation is end to end test. When you debug, you can find `inOperationCoercion` return false, but tests can still pass cuz `ComparableOperandTypeChecker#checkOperandTypes` passes.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java [hunk 1]
```diff
@@ -535,7 +535,7 @@ public int size() {
           coerced = rowTypeCoercion(scope1, node2, i, desired) || coerced;
         }
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:536]: Add this `coerced = coerced || listCoerced;` code in line 530 is enough, can we try to give a test case here ?

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java:536]: I'll give it a try.

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java]: > Add this `coerced = coerced || listCoerced;` code in line 530 is enough, can we try to give a test case here ?

Hi Danny, I've updated my PR and add a test, would you take a look?

**Comment by zinking** [comment]
LGTM

File: core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java [hunk 2]
```diff
@@ -53,6 +67,26 @@
...
+    assertFalse(isInCoercion("'1' in ('1', '2', '3')"));
+    assertTrue(isInCoercion("1 in ('1', '2', '3')"));
+    assertTrue(isInCoercion("1 not in ('1', '2', '3')"));
+  }
+
+  private boolean isInCoercion(String expr) throws SqlParseException {
+    final SqlCall inExpr = (SqlCall) SqlParser.create(expr).parseExpression();
+    final RelDataTypeFactory typeFactory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RelDataType charType = typeFactory.createSqlType(SqlTypeName.CHAR);
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:80]: After some dig into the code, i found that if the RHS is a node list, the common type is pre-computed for SQL in operator, and the nodes type was actually changed during the implicit type coercion, so the final validation passes and are still valid.

So it is hard to write a test, i would be ok if this patch does not have a test, the test is complex and it is hard to understand what it is testing.

Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:80]: > After some dig into the code, i found that if the RHS is a node list, the common type is pre-computed for SQL in operator, and the nodes type was actually changed during the implicit type coercion, so the final validation passes and are still valid.
> 
> So it is hard to write a test, i would be ok if this patch does not have a test, the test is complex and it is hard to understand what it is testing.

Yes, the fix only affects the return of "inOperationCoercion", the node type was updated during "inOperationCoercion", it is a side effect of this function.

IMO, having the test is better than no ut. On the one hand, ut puts constraints on the method, on the other hand, it is easy for debugging the method "inOperationCoercion".

Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/test/TypeCoercionConverterTest.java:80]: The test is not very complicated, it just needs a "SqlCallBinding".

**Pull request closed without merging**